### PR TITLE
Make object model consistent with Java, Go, & Node

### PR DESCRIPTION
**Pull request by abhinav**
The general object model is outlined below. This is similar to what Java and Go
use.

RPC
---

    RPC
        start()
        stop()

        register(Registrant)
        channel(ServiceName) -> Channel

        handle(Request<bytes>) -> Future<Response<bytes>>

Transport
---------

    Outbound
        call(Request<bytes>) -> Response<bytes>

    Inbound
        start(Handler<bytes, bytes>)
        stop()

    Handler
        handle(Request<bytes>) -> Response<bytes>

    Channel(Outbound)
        call(Request<bytes>) -> Response<bytes>

    Registrant
        procedures map[string]Handler<bytes, bytes>

The HTTP and TChannel encodings provide Inbound and Outbound implementations.

Encoding
--------

    OutgoingEncoder<I, O>
        encode_request(Request<I>) -> Request<bytes>
        decode_response(Response<bytes>) -> Response<O>

    IncomingEncoder<I, O>
        decode_request(Request<bytes>) -> Request<I>
        encode_response(Response<O>) -> Response<bytes>
        encode_exc_info(exc_info) -> Response<bytes>

On top of the Encoders, we have,

    AsyncClient(Channel)
        call(OutgoingEncoder<I, O>, Request<I>) -> Future<Response<O>>

    AsyncHandler(Handler<I, O>)
        handle(IncomingEncoder<I, O>, Request<bytes>) -> Future<Response<bytes>>

    procedure(Name, Handler<I, O>) -> Registrant

The Raw, JSON, and Thrift encodings provide specialized versions of
`AsyncClient` and `AsyncHandler`, and implement their own `procedure`
decorators.

### CODE REVIEW

**Comment by abhinav** [comment]
@yarpc/python @yarpc/yarpc

File: yarpc/transport/transport.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        encoding=None,
+        ttl=None,
+        procedure=None,
+        headers=None,
+        body=None,
+    ):
+        self.caller = caller
+        self.service = service
+        self.encoding = encoding
+        self.ttl = ttl  # TODO what unit is this
```
Comment (kriskowal) [inline yarpc/transport/transport.py]: I believe we resolved to use timedeltas for the python internal representation.

**Comment by kriskowal** [comment]
Do start and stop return futures?

For codecs:

s/Incoming/Inbound/
s/Outgoing/Outbound/

**Comment by abhinav** [comment]
Start and stop do not return futures. They simply block until we're ready to accept connections/have stopped listening.

Good call on codec naming.

**Comment by breerly** [comment]
@abhinav sorry, you'll need to rebase ;(

File: tests/conftest.py [hunk 2]
```diff
@@ -40,24 +66,36 @@ def transport_name(request):
...
+        service='test-service',
+        inbounds=[inbound],
+        outbounds={
+            'test-service': outbound,
+        },
+    )
+
+    try:
+        rpc.start()
+        yield rpc
```
Comment (breerly) [inline tests/conftest.py]: whats with the `yield rpc` here? is this something customers will have to do?

File: tests/yarpc/encoding/test_json.py [hunk 3]
```diff
@@ -24,76 +24,60 @@
...
+    resp = hi(req)
+    assert isinstance(resp, Response)
+    assert resp.body == {'some': 'data'}
 
-    rpc.json.register('hi', hi)
-    rpc.start()
 
-    resp = yield rpc.json(
-        service='test-service',
+@pytest.mark.xfail
```
Comment (breerly) [inline tests/yarpc/encoding/test_json.py]: fix in different diff? or you gonna line this up here?

File: yarpc/dispatcher.py [hunk 4]
```diff
@@ -1,41 +0,0 @@
...
-# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
-# THE SOFTWARE.
-
-from __future__ import (
-    absolute_import, division, print_function, unicode_literals
-)
-
-
-class Dispatcher(object):
```
Comment (breerly) [inline yarpc/dispatcher.py]: yay, goodbye dispatcher :)

File: yarpc/encoding/__init__.py [hunk 5]
```diff
@@ -21,12 +21,3 @@
 from __future__ import (
     absolute_import, division, print_function, unicode_literals
 )
-
-
-RAW = 'raw'
-JSON = 'json'
-THRIFT = 'thrift'
```
Comment (breerly) [inline yarpc/encoding/__init__.py]: yeah, this glossary didnt make sense considering it made yarpc kind of want to know about all possible encodings

File: yarpc/encoding/json.py [hunk 6]
```diff
@@ -18,69 +18,105 @@
...
+            procedure='hello',
+            headers={'token': 'foo'},
+            body={'name': 'somecaller'},
+        ))
+    """
+
+    __slots__ = ()
+
+    def call(self, request):
+        return super(JSONClient, self).call(JSONEncoder, request)
```
Comment (breerly) [inline yarpc/encoding/json.py]: I'm not wild about the use of inheritance here

File: yarpc/encoding/thrift.py [hunk 7]
```diff
@@ -18,138 +18,271 @@
...
+# TODO perhaps we should introduce a Thrift-only Request type which makes it
+# impossible to set the procedure.
+
+
+def load(path):
+    """Loads the Thrift file at the given path."""
+    return wrap(thriftrw.load(path))
+
+
+def wrap(thriftrw_module):
```
Comment (breerly) [inline yarpc/encoding/thrift.py]: suggest `_wrap` - i dont want to expose this in the public api

File: yarpc/encoding/thrift.py [hunk 8]
```diff
@@ -18,138 +18,271 @@
...
+        setattr(module, n, c)
+
+    # Attach includes
+    for m in thriftrw_module.__includes__:
+        setattr(module, m.__name__, m)
+
+    return module
+
+
+class Service(object):
```
Comment (breerly) [inline yarpc/encoding/thrift.py]: suggest `_Service`

File: yarpc/encoding/thrift.py [hunk 9]
```diff
@@ -18,138 +18,271 @@
...
+    def name(self):
+        return self.service_spec.name
+
+    def __str__(self):
+        return 'Service(%s)' % self.name
+
+    __repr__ = __str__
+
+
+class Function(object):
```
Comment (breerly) [inline yarpc/encoding/thrift.py]: suggest `_Function`

File: yarpc/encoding/thrift.py [hunk 10]
```diff
@@ -18,138 +18,271 @@
...
+
+class ThriftClient(AsyncClient):
+    """ThriftClient builds a client that sends Thrift requests via the given
+    channel.
+
+    .. code-block:: python
 
-class ThriftEncoding(object):
+        foo = thrift.load('foo.thrift')
+        client = thrift.ThriftClient(rpc.channel('foo'))
```
Comment (breerly) [inline yarpc/encoding/thrift.py]: this example is looking really crisp :+1:

File: yarpc/encoding/thrift.py [hunk 11]
```diff
@@ -18,138 +18,271 @@
...
-        rpc_call_future.add_done_callback(rpc_call_done)
+    def handle(self, request):
+        encoder = _IncomingEncoder(self.function_spec)
+        return super(ThriftHandler, self).handle(encoder, request)
 
-        return return_future
 
-    def register(self, thrift_service, handler):
-        # TODO enable calling with decorator
+class ThriftProcedure(object):
```
Comment (breerly) [inline yarpc/encoding/thrift.py]: suggest `_ThriftProcedure`

File: yarpc/encoding/thrift.py [hunk 12]
```diff
@@ -18,138 +18,271 @@
...
+
+        if response_body is not None:
+            response_body = self.module.dumps(response_body)
+            return Response(body=response_body)
+
+        # No match. Re-raise.
+        raise exc_info[0], exc_info[1], exc_info[2]
+
+
+class ThriftHandler(AsyncHandler):
```
Comment (breerly) [inline yarpc/encoding/thrift.py]: suggest `_ThriftHandler`

File: yarpc/transport/transport.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+Types defined in this file MUST be agnostic of whether we're synchronous or
+asynchornous.
+"""
+
+from __future__ import (
+    absolute_import, division, print_function, unicode_literals
+)
+
+
+class Request(object):
```
Comment (breerly) [inline yarpc/transport/transport.py]: I feel like these should stay top-level, the req/res objects are not transport-specific

File: yarpc/transport/transport.py (hunk 1 above)
Comment (breerly) [inline yarpc/transport/transport.py]: lets do that in another PR. issue here - https://github.com/yarpc/yarpc-python/issues/22

**Comment by breerly** [comment]
Really pumped with the direction of this - let's get those tests passing and I'm a big :+1:

File: tests/yarpc/encoding/test_json.py (hunk 3 above)
Comment (abhinav) [inline tests/yarpc/encoding/test_json.py]: Different diff

File: tests/conftest.py (hunk 2 above)
Comment (abhinav) [inline tests/conftest.py]: No. This is how the `yield_fixture` API works. It's so that I can perform cleanup (`rpc.stop()`) after the test.

File: yarpc/encoding/json.py (hunk 6 above)
Comment (abhinav) [inline yarpc/encoding/json.py]: True. I wasn't sure about this either. Removed in favor of wrapping the `AsyncClient` and `AsyncHandler`.

Comment (abhinav) [inline yarpc/encoding/json.py]: Also dropped from sync client implementation.

File: yarpc/encoding/thrift.py (hunk 7 above)
Comment (abhinav) [inline yarpc/encoding/thrift.py]: I actually wouldn't be opposed to exposing this in a public API for cases when people do `thriftrw.load` themselves. I'll hide it for now and we can expose it if necessary.

File: yarpc/encoding/thrift.py (hunk 8 above)
Comment (abhinav) [inline yarpc/encoding/thrift.py]: I think we'll want most of the `thrift.load(..)` logic to be in a non-public submodule of `yarpc.encoding.thrift`. No reason to rename the classes to be non-public if the module itself will be non-public.

Comment (abhinav) [inline yarpc/encoding/thrift.py]: I'll add a TODO to move things into a submodule.

File: yarpc/encoding/thrift.py (hunk 9 above)
Comment (abhinav) [inline yarpc/encoding/thrift.py]: Addressed above

File: yarpc/transport/transport.py (hunk 13 above)
Comment (abhinav) [inline yarpc/transport/transport.py]: These types are low-level Request and Response types. They are not specific to a single transport. They're specific to the low-level Transport abstraction. We re-export them from the yarpc module.

File: yarpc/encoding/thrift.py [hunk 14]
```diff
@@ -18,138 +18,289 @@
...
-                thrift_response = thrift_module.loads(
-                    thrift_request.result_type,
-                    result.body,
-                )
-            result.body = _unwrap_body(
-                response_spec=thrift_request.result_type.type_spec,
-                body=thrift_response,
-                procedure=request.procedure
+        request.body = self.module.dumps(request.body)
+        return request
```
Comment (blampe) [inline yarpc/encoding/thrift.py]: I love that this decouples serialization from sending the req over the wire.

**Comment by abhinav** [comment]
@blampe @breerly @junchaowu Going to merge this if no objections

File: tests/yarpc/encoding/test_json.py (hunk 3 above)
Comment (abhinav) [inline tests/yarpc/encoding/test_json.py]: Created https://github.com/yarpc/yarpc-python/issues/70

**Comment by breerly** [comment]
:shipit:
